src/main/java/ch/njol/skript/expressions/ExprArgument.java

```diff
@@ -44,148 +46,246 @@
-				}
-				arg = currentArguments.get(i - 1);
+				what = ORDINAL;
 				break;
 			case 3:
-				if (currentArguments.size() == 1) {
-					arg = currentArguments.get(0);
-				} else {
-					Skript.error("'argument(s)' cannot be used if the command has multiple arguments. Use 'argument 1', 'argument 2', etc. instead", ErrorQuality.SEMANTIC_ERROR);
-					return false;
-				}
+				what = parseResult.mark == 1 ? ALL : SINGLE;
 				break;
 			case 4:
 			case 5:
-				@SuppressWarnings("unchecked")
-				final ClassInfo<?> c = ((Literal<ClassInfo<?>>) exprs[0]).getSingle();
-				@SuppressWarnings("null")
-				final int num = parser.regexes.size() > 0 ? Utils.parseInt(parser.regexes.get(0).group()) : -1;
-				int j = 1;
-				for (final Argument<?> a : currentArguments) {
-					if (!c.getC().isAssignableFrom(a.getType()))
-						continue;
-					if (arg != null) {
-						Skript.error("There are multiple " + c + " arguments in this command", ErrorQuality.SEMANTIC_ERROR);
+				what = CLASSINFO;
+				break;
+			default:
+				assert false;
+		}
+
+		if (!scriptCommand && what == CLASSINFO) {
+			Skript.error("Command event arguments are strings, meaning type specification is useless");
+			return false;
+		}
+
+		List<Argument<?>> currentArguments = Commands.currentArguments;
+		if (scriptCommand && (currentArguments == null || currentArguments.isEmpty())) {
+			Skript.error("This command doesn't have any arguments", ErrorQuality.SEMANTIC_ERROR);
+			return false;
+		}
+
+		if (what == ORDINAL) {
+			// Figure out in which format (1st, 2nd, 3rd, Nth) argument was given in
+			MatchResult regex = parseResult.regexes.get(0);
+			String argMatch = null;
+			for (int i = 1; i <= 4; i++) {
+				argMatch = regex.group(i);
+				if (argMatch != null) {
+					break; // Found format
+				}
+			}
+			assert argMatch != null;
+			ordinal = Utils.parseInt(argMatch);
+			if (scriptCommand && ordinal > currentArguments.size()) { // Only check if it's a script command as we know nothing of command event arguments
+				Skript.error("This command doesn't have a " + StringUtils.fancyOrderNumber(ordinal) + " argument", ErrorQuality.SEMANTIC_ERROR);
+				return false;
+			}
+		}
+
+		if (scriptCommand) { // Handle before execution
+			switch (what) {
+				case LAST:
+					argument = currentArguments.get(currentArguments.size() - 1);
+					break;
+				case ORDINAL:
+					argument = currentArguments.get(ordinal - 1);
+					break;
+				case SINGLE:
+					if (currentArguments.size() == 1) {
+						argument = currentArguments.get(0);
+					} else {
+						Skript.error("This command has multiple arguments, meaning it is not possible to get the 'argument'. Use 'argument 1', 'argument 2', etc. instead", ErrorQuality.SEMANTIC_ERROR);
 						return false;
 					}
-					if (j < num) {
-						j++;
-						continue;
+					break;
+				case ALL:
+					Skript.error("'arguments' cannot be used for script commands. Use 'argument 1', 'argument 2', etc. instead", ErrorQuality.SEMANTIC_ERROR);
```

Comment:
This might be better to revisit later due to the awkwardness of the argument system (they are stored in a static list right now). I'm not entirely sure about it either as I'm not sure the returned list (of random objects) would have too much value to a user (although users do like weird things).